jetty-server/src/test/java/org/eclipse/jetty/server/DetectorConnectionFactoryTest.java

```diff
@@ -0,0 +1,584 @@
+                pw.close();
+                return byteBuffer + "\n" + sw.toString();
+            }
+        }
+    }
+
+    @AfterEach
+    public void destroy() throws Exception
+    {
+        if (_server != null)
+            _server.stop();
+        if (_assertingByteBufferPool != null)
+            _assertingByteBufferPool.assertAllBuffersWereReleased();
+    }
+
+    private void start(ConnectionFactory... connectionFactories) throws Exception
+    {
+        _server = new Server();
+        _assertingByteBufferPool = new AssertingByteBufferPool(100);
+        _server.addBean(_assertingByteBufferPool);
+        _server.addConnector(new ServerConnector(_server, 1, 1, connectionFactories));
+        _server.setHandler(new DumpHandler());
+        _server.start();
+    }
+
+    private String getResponse(String request) throws Exception
+    {
+        try (Socket socket = new Socket(_server.getURI().getHost(), _server.getURI().getPort()))
+        {
+            socket.getOutputStream().write(request.getBytes(StandardCharsets.US_ASCII));
+
+            StringBuilder sb = new StringBuilder();
+            BufferedReader reader = new BufferedReader(new InputStreamReader(socket.getInputStream(), StandardCharsets.US_ASCII));
+
+            while (true)
+            {
+                String line = reader.readLine();
+                if (line == null)
+                    break;
+                sb.append(line).append("\n");
+            }
+
+            return sb.length() == 0 ? null : sb.toString();
+        }
+    }
+
+    private String getResponseOverSsl(String request) throws Exception
+    {
+        String keystore = MavenTestingUtils.getTestResourceFile("keystore").getAbsolutePath();
+        SslContextFactory sslContextFactory = new SslContextFactory.Server();
+        sslContextFactory.setKeyStorePath(keystore);
+        sslContextFactory.setKeyStorePassword("storepwd");
+        sslContextFactory.setKeyManagerPassword("keypwd");
+        sslContextFactory.start();
+
+        SSLSocketFactory socketFactory = sslContextFactory.getSslContext().getSocketFactory();
+        try (Socket socket = socketFactory.createSocket(_server.getURI().getHost(), _server.getURI().getPort()))
+        {
+            socket.getOutputStream().write(request.getBytes(StandardCharsets.US_ASCII));
+
+            StringBuilder sb = new StringBuilder();
+            BufferedReader reader = new BufferedReader(new InputStreamReader(socket.getInputStream(), StandardCharsets.US_ASCII));
+
+            while (true)
+            {
+                String line = reader.readLine();
+                if (line == null)
+                    break;
+                sb.append(line).append("\n");
+            }
+
+            return sb.length() == 0 ? null : sb.toString();
+        }
+        finally
+        {
+            sslContextFactory.stop();
+        }
+    }
+
+    @Test
+    public void testDetectingSslProxyToHttpNoSslNoProxy() throws Exception
```

Comment:
Better name would be `testDetectionUnsuccessfulUpgradesToFirstAfter()`?

Reply:
I changed this test's name to `testDetectionUnsuccessfulUpgradesToNextProtocol`.